ActivatorUtilities not depending on ctor order for creating instances

maryamariyan opened the pull request:
Went over existing test cases and the ones presented in PR https://github.com/dotnet/runtime/pull/67493 by @mapogolions and adding some more use cases to investigate behavioral change with this fix.

Fixes #46132

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -350,30 +360,25 @@ public void GetServiceOrCreateInstanceUnregisteredService()
         }
 
-        [Fact]
-        public void CreateInstance_WithAbstractTypeAndPublicConstructor_ThrowsCorrectException()
```
Why is this test being removed?

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -251,8 +251,13 @@ public void TypeActivatorUsesMarkedConstructor(CreateInstanceFunc createFunc)
         [MemberData(nameof(CreateInstanceFuncs))]
         public void TypeActivatorThrowsOnMultipleMarkedCtors(CreateInstanceFunc createFunc)
         {
+            // Arrange
```
Do these 2 tests need to change?

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -32,17 +32,19 @@ public static class ActivatorUtilities
             params object[] parameters)
         {
             int bestLength = -1;
+            int largestLengthWithMultipleAcceptedCtors = -1;
```
I'm not sure I get what this variable is trying to track based on its name.

Maybe it would be more clear if it was `bool multipleBestLengthFound = false`.

And then set it to `true` / `false` as appropriate below.

maryamariyan replied:
I didn't want it to be bool, because it only matters for us to throw ambiguous exception when for the best length (largest found length) we've found multiple ctors.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -73,7 +79,13 @@ public static class ActivatorUtilities
-            return bestMatcher.CreateInstance(provider);
+            if (largestLengthWithMultipleAcceptedCtors == bestLength)
+            {
+                string? message = $"multiple constructors for type '{instanceType}' were found with length {bestLength}.";
```
I know existing strings are hard-coded, but let's put new strings into the resx file.

maryamariyan replied:
I'll cover that separately when also updating the package versions for `Autofac`, etc.

eerhardt replied:
Why? You are adding new strings now - why add the debt?

maryamariyan replied:
because I wanted to update versions in a separate PR. now it's addressed in this same PR

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -73,7 +79,13 @@ public static class ActivatorUtilities
-            return bestMatcher.CreateInstance(provider);
+            if (largestLengthWithMultipleAcceptedCtors == bestLength)
+            {
+                string? message = $"multiple constructors for type '{instanceType}' were found with length {bestLength}.";
```
```suggestion
                string? message = $"Multiple constructors for type '{instanceType}' were found with length {bestLength}.";
```

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -73,7 +79,13 @@ public static class ActivatorUtilities
+                throw new InvalidOperationException(message);
+            }
+
+            return bestMatcher.CreateInstance(provider!);
```
Why is this `!` valid? I think it would be better if we either threw an `ArgumentNullException` or actually made the parameter nullable.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -350,15 +362,12 @@ public int Match(object[] givenParameters)
                         if (_parameterValues[applyIndex] == null &&
                             _parameters[applyIndex].ParameterType.IsAssignableFrom(givenType))
                         {
+                            applyExactLength++;
```
It might make sense to rename `applyExactLength` to a better name now. I don't think that matches what it is used for now.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Autofac.cs:
```diff
@@ -9,12 +9,20 @@ namespace Microsoft.Extensions.DependencyInjection.Specification
             "ScopesAreFlatNotHierarchical",
-            "ServiceScopeFactoryIsSingleton"
+            "ServiceScopeFactoryIsSingleton",
+            // different behavior between net7 and net462
```
I don't think we should be skipping these tests, especially on net7.

If we think skipping these tests is the right thing - we should only do it on net462.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -31,18 +31,25 @@ public static class ActivatorUtilities
             [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicConstructors)] Type instanceType,
             params object[] parameters)
         {
+            if (provider == null)
```
Is there a test that ensures this throws an `ArgumentNullException`?

maryamariyan replied:
Yes, `TypeActivatorThrowsOnNullProvider`. Moved it over to the right `ActivatorUtilitiesTests` file.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -214,9 +221,9 @@ public void TypeActivatorCreateFactoryDoesNotAllowForAmbiguousConstructorMatches
-        public void TypeActivatorCreateInstanceUsesFirstMathchedConstructor(object value, string ctor)
+        [InlineData("", "IFakeService, string", "string")]
+        [InlineData(5, "IFakeService, int", "IFakeService, int")]
+        public void TypeActivatorCreateInstanceUsesLongestAvailableConstructor(object value, string ctor, string oldBehavior)
```
`oldBehavior` is not a good name to choose. "Old" as relative to what? It quickly becomes ambiguous.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -247,12 +261,23 @@ public void TypeActivatorUsesMarkedConstructor(CreateInstanceFunc createFunc)
+        [Fact]
+        public void TypeActivatorThrowsOnNullProvider()
+        {
+            var exception = Assert.Throws<ArgumentNullException>(() => CreateInstance<ClassWithMultipleMarkedCtors>(CreateInstanceDirectly, null, "hello"));
```
This test shouldn't be in the Specification tests, but instead in the DI.Tests\ActivatorUtilitiesTests.cs.

The Specification tests are to ensure the external DI implementations meet the minimum expected behavior. Things under `DI.Tests` ensure our behavior is correct.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -247,12 +261,23 @@ public void TypeActivatorUsesMarkedConstructor(CreateInstanceFunc createFunc)
         {
+            // Arrange
+            var serviceCollection = new TestServiceCollection();
+            serviceCollection.AddSingleton<IFakeService, FakeService>();
```
Do we need this service to be registered? I think it is irrelevant.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -262,8 +287,13 @@ public void TypeActivatorThrowsOnMultipleMarkedCtors(CreateInstanceFunc createFu
         {
+            // Arrange
+            var serviceCollection = new TestServiceCollection();
+            serviceCollection.AddSingleton<IFakeService, FakeService>();
```
Same as above - there shouldn't need to be any services registered.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Autofac.cs:
```diff
@@ -9,8 +9,6 @@ namespace Microsoft.Extensions.DependencyInjection.Specification
 {
     public class AutofacDependencyInjectionSpecificationTests : SkippableDependencyInjectionSpecificationTests
     {
-        public override bool SupportsIServiceProviderIsService => false;
```
Can we get rid of the `SupportsIServiceProviderIsService` boolean all together now that all the External DI implementations support it?

maryamariyan replied:
no because `Lamar` is broken for example

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Microsoft.Extensions.DependencyInjection.ExternalContainers.Tests.csproj:
```diff
@@ -15,11 +15,11 @@
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Specification.Tests\src\Microsoft.Extensions.DependencyInjection.Specification.Tests.csproj" />
-    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="7.2.0" />
+    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="8.0.0" />
     <PackageReference Include="DryIoc.Microsoft.DependencyInjection" Version="5.1.0" />
```
Can we update all these versions to the latest?

maryamariyan replied:
they're all updated to latest

maryamariyan commented:
Lamar seems broken.

Code:
https://github.com/JasperFx/lamar/blob/v8.0.0/src/Lamar.Microsoft.DependencyInjection/HostBuilderExtensions.cs#L37-L53

It's not properly setting `IServiceProviderIsService`. Therefore I'm gonna keep it disabled in my test.

halter73 commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -31,18 +31,25 @@ public static class ActivatorUtilities
             bool seenPreferred = false;
 
             ConstructorMatcher bestMatcher = default;
+            IServiceProviderIsService? serviceProviderIsService = provider.GetService<IServiceProviderIsService>();
```
We should be careful about relying on an IServiceProviderIsService implementation giving false negatives. This should never happen with the built-in ServiceProvider implementation, but it could happen with third part containers. This turned out to be a major point of contention when [approving the IServiceProviderIsService API](https://github.com/dotnet/runtime/issues/53919#issuecomment-858861117). False positives should not happen with any container.

[Here's a timestamped video of that discussion.](https://www.youtube.com/watch?v=YRcGjsFbx4Y&t=0h55m13s)

I think there should at least be a way for developers to force ActivatorUtilities to use a specific constructor even if an IServiceProviderIsService exists and `IServiceProviderIsService.IsService(parameterType)` returns false. In ASP.NET Core for example, if you use the `[FromServices]` attribute, we will try to resolve the service regardless of what the IServiceProviderIsService has to say about it.

halter73 replied:
@tillig Do you have any thoughts on this?

halter73 replied:
I forgot ActivatorUtilitiesConstructorAttribute exists because we so rarely use it, but maybe that sufficient for this case. And this only affects ties, right? I'll mark this as resolved.

tillig replied:
Agreed, `IsService` is a "best effort" API, not a guarantee. In Autofac, for example, registrations may come from a dynamic registration source rather than be added explicitly to the container, in which case [at least the Autofac implementation for `IsService` might return `false`](https://github.com/autofac/Autofac.Extensions.DependencyInjection/blob/bf61f9225b078359c7e9e688b3823990d056a4f2/src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs#L53) even if the target could technically be resolved. It will _never_ return `true` if it can't be resolved.

tillig replied:
IIRC there was also some discussion about the `IsService` being something that indicates "it's registered" but _not_ that it can necessarily be successfully resolved. All that said, I just did a bunch of testing for the 90% case just now to see what we detect as registered vs. not and we're _usually_ correct and do detect most of the dynamic stuff, too. I know @alistairjevans did a lot of work to help with assembly trimming and may have impacted our accuracy in a positive way here. Regardless, I think you're right that it's likely not a big deal since it is such an edge case.

alistairjevans replied:
Dynamic sources generally should be ok
..the false negative problem came from the fact that, in Autofac (and a bunch of other IoC containers) you can register service implementations in a nested unit-of-work/request scope, that didn't exist in the root scope.

So, if `IsService` is called from the root scope, and returns false, it could conceivably return true if called from a more nested scope.

For perfect accuracy of checking whether a given method argument is a service, the `IsService` call would need to happen on the same scope the services are about to be resolved from.

eerhardt replied:
> For perfect accuracy of checking whether a given method argument is a service, the IsService call would need to happen on the same scope the services are about to be resolved from.

In this case it is. The check for `IsService` happens when trying to select the constructor. Then once a ctor is selected, the same `IServiceProvider` is used to get the services to populate the ctor's arguments.

tillig commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.External.Tests/Microsoft.Extensions.DependencyInjection.ExternalContainers.Tests.csproj:
```diff
@@ -15,16 +15,16 @@
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Abstractions\src\Microsoft.Extensions.DependencyInjection.Abstractions.csproj" SkipUseReferenceAssembly="true" />
     <ProjectReference Include="$(LibrariesProjectRoot)Microsoft.Extensions.DependencyInjection.Specification.Tests\src\Microsoft.Extensions.DependencyInjection.Specification.Tests.csproj" />
-    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="7.2.0" />
+    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="8.0.0-beta133" />
```
```suggestion
    <PackageReference Include="Autofac.Extensions.DependencyInjection" Version="8.0.0" />
```

v8 was released.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ActivatorUtilitiesTests.cs:
```diff
@@ -0,0 +1,190 @@
+            services.AddSingleton(b);
+            using var provider = services.BuildServiceProvider();
+
+            Assert.Throws<InvalidOperationException>(() =>
```
(nit) Do we want to verify the exception message here?

eerhardt reviewed (dismissed):
This LGTM. Thanks for the good work here @maryamariyan!

We can update the other external providers in a separate PR.

* Lamar once their bug is fixed.
* Grace once a new stable version is released
* DryIoc should be able to be updated now, and remove the SupportsIServiceProviderIsService  override, but that can be a separate PR.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -31,18 +31,25 @@ public static class ActivatorUtilities
                 {
                     var matcher = new ConstructorMatcher(constructor);
                     bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
+                    int length = matcher.Match(parameters, serviceProviderIsService);
```
Maybe one thing that should be changed here is what @halter73 mentioned:

> I think there should at least be a way for developers to force ActivatorUtilities to use a specific constructor even if an IServiceProviderIsService exists and IServiceProviderIsService.IsService(parameterType) returns false

With the old code, if a ctor has `ActivatorUtilitiesConstructorAttribute` on it and it takes all the `parameters` supplied to `CreateInstance`, we will use that ctor.

With the new code, if a ctor has `ActivatorUtilitiesConstructorAttribute` on it and takes all the `parameters` supplied to `CreateInstance`, but the `IServiceProviderIsService` gives a false negative on one of the IsService calls for a different parameter, we will not use the ctor but throw an exception.

So maybe we need to pass `isPreferred` into `matcher.Match`, and only check the supplied `parameters` when `isPreferred=true`.

halter73 replied:
I also think it's important for CreateInstance to fall back to something closer to the older logic (without depending on ctor order) if IServiceProviderIsService would cause no constructor to be matched.

When we're activating a type the customer does not control without `[ActivatorUtilitiesConstructor]` like our built-in middleware, we don't want that to start breaking code that worked before just because someone is using a third-party container or mocked an incomplete `IServiceProviderIsService` in their test.

Ideally, CreateInstance would only consult IServiceProviderIsService in situations where CreateFactory would always fail and behave the same otherwise.

eerhardt replied:
> I also think it's important for CreateInstance to fall back to something closer to the older logic (without depending on ctor order) if IServiceProviderIsService would cause no constructor to be matched.

Can you describe the rules for "something closer to the older logic (without depending on ctor order)"? 

The older logic, AFAICT, picks the constructor that:

1. matched all the provided parameters
2. has a matching parameter in the exact same position in its argument list
3. and is highest matching position (furthest to the end of the parameter list)
4. If multiple ctors meet that criteria, the first was chosen.

The logic came in https://github.com/aspnet/DependencyInjection/pull/14, but there are no comments that describe the intended behavior / rules.

maryamariyan commented:
The pull request is ready for review.

changes were made based on the agreements in comment https://github.com/dotnet/runtime/issues/46132#issuecomment-1267749902:

- Goal: We want to completely fix the bug with behaviour dependence on order of constructor definition.
- All the instance arguments passed in to `CreateInstance` were required before and should remain required after the breaking change.
- Generally we do not want `CreateInstance` to be as strict as `CreateFactory`.
- `CreateInstance` needs to account for the case where a container is potentially misconfigured. If for any reason `IServiceProviderIsService` ends returns false for a service which is actually registered, should still work today.

For example, given:
```c#
public class A
{
   A(B b) { }
}
```
`ActivatorUtilities.CreateInstance(provider, typeof(A))` should be able to properly instantiate `A` if `B` is registered. Even if `IServiceProviderIsService` is not available or `IServiceProviderIsService.IsService` for `B` returns false for some reason incorrectly.

General idea:
- Try to find the longest constructor matching all parameters based on the behaviour of `IServiceProviderIsService`.
- If none are found, try to fallback to `CreateFactory`, more than one throw
- if `IServiceProviderIsService` is not present fallback to `CreateFactory`. 
- If `IServiceProviderIsService` is configured wrong, or does not exist, it is OK for the API to function incorrectly or ambiguously.

maryamariyan commented:
note for reviewers: view diff per commit for improved readability

maryamariyan commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -84,8 +114,11 @@ public static class ActivatorUtilities
             if (bestLength == -1)
             {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
```
falling back to CreateFactory logic here

maryamariyan commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -42,18 +42,48 @@ public static class ActivatorUtilities
+                // second check is to ignore finding longest ctor if container is setup incorrectly
+                !serviceProviderIsService.IsService(typeof(IServiceProviderIsService)))
+            {
+                FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo? constructorInfo, out int?[]? parameterMap);
```
falling back to CreateFactory logic here

maryamariyan commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -310,25 +343,21 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
         // Creates an injective parameterMap from givenParameterTypes to assignable constructorParameters.
         // Returns true if each given parameter type is assignable to a unique; otherwise, false.
-        private static bool TryCreateParameterMap(ParameterInfo[] constructorParameters, Type[] argumentTypes, out int?[] parameterMap)
+        private static int TryCreateParameterMap(ParameterInfo[] constructorParameters, Type[] argumentTypes, out int?[] parameterMap)
```
sharing the same logic here for both CreateInstance and CreateFactory

maryamariyan commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -357,65 +386,45 @@ public ConstructorMatcher(ConstructorInfo constructor)
-                    return matchedParameterLength;
-                }
-
+            public int MatchAll(int requiredParameterLength, IServiceProviderIsService serviceProviderIsService)
```
`MatchAll` to get longest ctor using `IServiceProviderIsService`

maryamariyan committed:
Fixes the bug with behaviour dependence on order of constructor definition.
Tries to find the longest constructor matching all parameters based on the behaviour of IServiceProviderIsService.

- All the instance arguments passed in to CreateInstance were required before and should remain required after the breaking change.
- Generally we do not want CreateInstance to be as strict as CreateFactory.
- CreateInstance needs to account for the case where a container is potentially misconfigured. If for any reason IServiceProviderIsService ends returns false for a service which is actually registered, should still work today, but it is ok to give false negative.

- If none are found, try to fallback to CreateFactory, more than one throw
- if IServiceProviderIsService is not present fallback to CreateFactory.
- If IServiceProviderIsService is configured wrong, or does not exist, it is OK for the API to function incorrectly or ambiguously.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -262,11 +286,19 @@ public void TypeActivatorThrowsOnMultipleMarkedCtors(CreateInstanceFunc createFu
+            Assert.Equal(message, exception.Message);
+
+            exception = Assert.Throws<InvalidOperationException>(() =>
+                CreateInstance<ClassWithMultipleAmbiguousCtorsAndAttribute>(createFunc, serviceProvider, 0, "hello"));
```
What does `ClassWithMultipleAmbiguousCtorsAndAttribute` mean? How is it different from `ClassWithAmbiguousCtorsAndAttribute`?

maryamariyan replied:
Removed, they seemed similar for the purpose of the test

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -61,19 +79,37 @@ public static class ActivatorUtilities
             }
 
-            return bestMatcher.CreateInstance(provider);
+            FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructorInfo, out int?[] parameterMap);
```
Do we have unit tests that cover the cases mentioned previously? Specifically the cases where `IServiceProviderIsService` returns `false` incorrectly.

maryamariyan replied:
added in https://github.com/dotnet/runtime/pull/75846/commits/6e46bd11ed915b24118055eb5a62d10754c00b18

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -212,12 +247,11 @@ private static MethodInfo GetMethodInfo<T>(Expression<T> expr)
 
-            matchingConstructor = constructorInfo;
-            matchingParameterMap = parameterMap;
+            matchingConstructor = constructorInfo!;
```
Are these `!` necessary? Both `TryFindPreferredConstructor` and `TryFindMatchingConstructor` are annotated as they won't return `null` parameters when returning `true`. So the compiler should be able to see that these will not be null.

maryamariyan replied:
I thought it would add to readability

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -31,18 +31,36 @@ public static class ActivatorUtilities
+            // if no such match is found we fallback to the same logic used by CreateFactory which would only allow creating an
+            // instance if all parameters given to CreateInstance only match with a single ctor
+            if (serviceProviderIsService != null &&
+                serviceProviderIsService.IsService(typeof(IServiceProviderIsService)))
```
I'm not sure this check is necessary. We should just need to check if `serviceProviderIsService` is null.

maryamariyan replied:
cc @halter73 do you have a scenario in which having both checks here would detect properly if a container is misconfigured?

halter73 replied:
I agree we should remove the `IsService(typeof(IServiceProviderIsService))` check. I don't think we can detect a misconfigured container ahead of time. It's just that if `IServiceProviderIsService` returns false so much that no constructors match, we should at least try the constructor `CreateFactory` would have called in the same situation assuming it would be unambiguous.

maryamariyan replied:
Removed the redundant check.

As for setting up a test for this case, I'm open to suggestions on how to set it up.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -31,18 +31,36 @@ public static class ActivatorUtilities
                     var matcher = new ConstructorMatcher(constructor);
                     bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
+                    int length = matcher.Match(parameters, serviceProviderIsService!);
```
```suggestion
                    int length = matcher.Match(parameters, serviceProviderIsService);
```

This has already been checked for `null`.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs:
```diff
@@ -61,19 +79,37 @@ public static class ActivatorUtilities
             {
-                string? message = $"A suitable constructor for type '{instanceType}' could not be located. Ensure the type is concrete and all parameters of a public constructor are either registered as services or passed as arguments. Also ensure no extraneous arguments are provided.";
-                throw new InvalidOperationException(message);
+                argumentTypes[i] = parameters[i].GetType();
```
Can one of the `parameters` be `null`? If so, this will NRE.

maryamariyan replied:
no parameters cannot be null on the CreateInstance signature:

https://github.com/dotnet/runtime/blob/2b07c8c46523b816b1eac5c8af092e52b20a69d2/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L29-L32

eerhardt replied:
We are checking for `null` elsewhere:

https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs#L345

Nothing is guaranteeing the caller doesn't pass `null` inside this array. The signature just tells the compiler to log a warning if they do. The caller can suppress that using `!`, or they could be using `netstandard2.0` which won't give them nullability warnings.

We shouldn't be throwing NullReferenceException here.

maryamariyan committed:
Apply PR feedback

maryamariyan committed:
Add mocked service

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ActivatorUtilitiesTests.cs:
```diff
@@ -0,0 +1,289 @@
+            mockedService.Setup(r => r.IsService(It.IsAny<Type>()))
+                .Returns(isService);
+            var services = new ServiceCollection();
+            services.AddScoped<IServiceProviderIsService>(p => mockedService.Object);
```
Does this actually work? Looking at the DI code, the `IServiceProviderIsService` service gets added after all the provided serviceDescriptors. Which I think means the built-in DI container's implementation will always win.

[https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L48-L53](https://github.com/dotnet/runtime/blob/3435b39194137676d9b2da12c8ebd21d69cd06b9/src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceProvider.cs#L48-L53)

maryamariyan committed:
mocked serviceprovider

maryamariyan committed:
Undo mocking fix NRE bug

maryamariyan committed:
Fix NRE for both CreateInstance and CreateFactory. moving test to other file

maryamariyan committed:
Remove unnecessary IsService check

maryamariyan committed:
Test badly formed container

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -23,7 +23,7 @@ private static object CreateInstanceDirectly(IServiceProvider provider, Type typ
         private static object CreateInstanceFromFactory(IServiceProvider provider, Type type, object[] args)
         {
-            var factory = ActivatorUtilities.CreateFactory(type, args.Select(a => a.GetType()).ToArray());
+            var factory = ActivatorUtilities.CreateFactory(type, args.Select(a => a?.GetType()).ToArray());
```
Why was this change necessary?

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -350,6 +378,39 @@ public void GetServiceOrCreateInstanceUnregisteredService()
         }
 
+        [Fact]
+        public void CreateInstance_BadlyConfiguredIServiceProviderIsService_ProperlyCreatesUnambiguousInstance()
```
I think this test is more appropriate in the `DI.Tests\ActivatorUtilitiesTests.cs` file. This is more about testing ActivatorUtilities, and not the DI Implementations.

Also - what happens when all the external providers SupportsIServiceProviderIsService=true? Now this test no longer does anything.

Instead, I think we should make a mock `IServiceProvider` that returns an `IServiceProviderIsService` that always returns `false`. And then use that to test the fallback in ActivatorUtilities.

eerhardt commented on src/libraries/Microsoft.Extensions.DependencyInjection.Specification.Tests/src/ActivatorUtilitiesTests.cs:
```diff
@@ -350,6 +378,39 @@ public void GetServiceOrCreateInstanceUnregisteredService()
+
+        [Theory]
+        [MemberData(nameof(CreateInstanceFuncs))]
+        public void CreateInstance_NullInstance_HandlesBadInputWithInvalidOperationException(CreateInstanceFunc createFunc)
```
This test should also be moved to DI.Tests\ActivatorUtilitiesTests.cs. It is testing ActivatorUtilities, not the external DI implementations.

tarekgh reviewed (approved):
LGTM!

Should we mark this PR as a breaking change to file a doc for it?

maryamariyan committed:
Update tests - Apply PR feedback

eerhardt reviewed (approved):
Looks great. Thanks for the great work here, @maryamariyan!
